### ISSUE TITLE
Feature Request: window.before(eventName, callback)

### ISSUE THREAD

**Issue by chaseWillden** [nwjs/nw.js#3099]
I wanted to suggest an idea sprung from [this issue](https://github.com/nwjs/nw.js/issues/2892).

Example usage:

```
var gui = require('nw.gui');
var win = gui.Window.get();
win.before('minimize', function(){
   console.log('minimized button was pressed');
});
win.before('maximize', function(){
   console.log('maximized button was pressed');
});
win.before('fullscreen', function(){
   console.log('We are about to go fullscreened');
});
```
Etc...

[References](https://github.com/nwjs/nw.js/blob/0afb5476f5cbd29d7efa299acedc28d86ae7cdd9/tests/manual_tests/window/popup.html)

**Comment by Wortex17** [comment]
Would this include an option "cancel" the event, just like the 'close' event?
Otherwise i would see not use in that, besides event-ordering (which IMO should not be done that way).

What would the effect look like? If the user maximizes the window by using shell interactions and the event is canceled, should it revert?

**Comment by chaseWillden** [comment]
@Wortex17 Yes I would assume this would include the same feature that a typical event has.

```
win.before('maximize', function(e){
   e.preventDefault();
   // or
   e.stopPropagation
});
```

**Comment by Wortex17** [comment]
this way of cancelling it would require making it an typical DOM event, not an node one. therefore, it would probably be best to send it from window, not nw's gui.window, which again would dissent the name 'before', as no maximize or minimize is fired on the DOM's window.
